File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        boolean dp[][] = new boolean[n][n];
        int start = 0,end = 0;
        
        for(int gap = 0;gap < n;gap++){
            for(int i = 0,j = gap;i< n && j < n;i++,j++){
                if(gap == 0){
                    dp[i][j] = true;
                }else if(gap == 1){
                    if(s.charAt(i) == s.charAt(j)){
                        dp[i][j] = true;
                    }else{
                        dp[i][j] = false;
                    }
                }else{
                    if(s.charAt(i) == s.charAt(j) && dp[i+1][j-1] == true){
                        dp[i][j] = true;
                    }else{
                        dp[i][j] = false;
                    }
                }
                if(dp[i][j] == true){
                    if(j-i > end - start){
                        start = i;
                        end = j;
                    }
                }
            }
        }
		
        return s.substring(start,end + 1);
    }
}

File: 5_ Longest_Palindromic_Substring.cpp
class Solution {
public:
    string longestPalindrome(string s)
    {
        string ans="";
        int len=0;
        for(int i=0;i<s.length();i++)
        {
            int lo,hi;
            lo=hi=i;
            while(lo>=0&&hi<s.length()&&(s[lo]==s[hi]))
                {
                    if((hi-lo+1)>len)
                    {
                        ans=s.substr(lo,hi-lo+1);
                        len=hi-lo+1;
                    }
                    lo--;
                    hi++;
                }
            lo=i,hi=i+1;
            while(lo>=0&&hi<s.length()&&s[lo]==s[hi])
            {
                if((hi-lo+1)>len)
                    {
                        ans=s.substr(lo,hi-lo+1);
                        len=hi-lo+1;
                    }
                    lo--;
                    hi++;
            }
        }
    return ans;    
    }   
};
